### PR TITLE
Wkt support staging

### PR DESCRIPTION
**Pull request by Bonkles**
This PR adds a new `wktPoly` param that very quickly gets basic POLYGON and MULTIPOLYGON support into Rapid.

### CODE REVIEW

File: modules/core/UrlHashSystem.js [hunk 1]
```diff
@@ -61,6 +61,8 @@ export class UrlHashSystem extends AbstractSystem {
 * __`id`__ - An OSM ID to select.
 * __`map`__ - A slash-separated `zoom/lat/lon/rot`.
 * __`offset`__ - Background imagery alignment offset in meters, formatted as `east,north`.
+* __`wkt`__ - Well-known text string to render as custom data.
```
Comment (Bonkles) [inline modules/core/UrlHashSystem.js:64]: Fix this!

File: modules/pixi/PixiLayerCustomData.js [hunk 2]
```diff
@@ -8,6 +8,7 @@ import { PixiFeatureLine } from './PixiFeatureLine.js';
 import { PixiFeaturePoint } from './PixiFeaturePoint.js';
 import { PixiFeaturePolygon } from './PixiFeaturePolygon.js';
 import { utilFetchResponse } from '../util/index.js';
+import { parse as wktParse } from 'wkt';
```
Comment (Bonkles) [inline modules/pixi/PixiLayerCustomData.js:11]: New module import that correctly interprets WKT strings of all kinds.

File: modules/pixi/PixiLayerCustomData.js [hunk 3]
```diff
@@ -96,11 +99,55 @@ export class PixiLayerCustomData extends AbstractLayer {
...
+
+    const wktStyle = { stroke: { width: 4, color: 0x0afaf, alpha: 0.5, cap: PIXI.LINE_CAP.ROUND }};
+    this.renderPolygons(frame, viewport, zoom, this._wktPolys, wktStyle);
+
+  }
+
+
+  /**
+   * createWktPolys
+   * creates WKT lines from a raw string supplied on the url (if specified) from param 'wkt'.
```
Comment (Bonkles) [inline modules/pixi/PixiLayerCustomData.js:119]: Fix this comment too.

File: modules/pixi/PixiLayerCustomData.js [hunk 4]
```diff
@@ -96,11 +99,55 @@ export class PixiLayerCustomData extends AbstractLayer {
...
+      console.error("Unable to parse wktPoly param");
+      return polys;
+    }
+
+
+    let newPoly = {
+      type: 'Feature',
+      geometry: {
+        type: parsedWkt.type,
+        coordinates: parsedWkt.coordinates, //wkt parses as a single polygon, so just return that.
```
Comment (Bonkles) [inline modules/pixi/PixiLayerCustomData.js:142]: Fix incorrect comment!

File: modules/pixi/PixiLayerCustomData.js [hunk 5]
```diff
@@ -206,7 +253,9 @@ export class PixiLayerCustomData extends AbstractLayer {
         if (feature.v !== version) {
           feature.v = version;
           feature.geometry.setCoords(coords);
-          feature.label = l10n.displayName(d.properties);
+          if (d.properties){
+            feature.label = l1n.displayName(d.properties);
```
Comment (Bonkles) [inline modules/pixi/PixiLayerCustomData.js:257]: @bhousel - I ran into stack traces while rendering this stuff, so I made the labelling contingent on whether properties existed (there are no labels supplied for these arbitrary polys).

**Review by Bonkles — COMMENTED** [review]
Left a number of comments to fix things now that I'm in 'review' mode.

File: modules/pixi/PixiLayerCustomData.js (hunk 5 above)
Comment (bhousel) [reply modules/pixi/PixiLayerCustomData.js:257]: Ok we can check.  I don't think wkt supports properties, but after it gets converted to GeoJSON it should have them.  I have seen some GeoJSON's in the wild where they are omitted, but you can always do something like `d.properties || {}` 

(also make sure it's `l10n` and not `l1n`)

File: modules/pixi/PixiLayerCustomData.js [hunk 6]
```diff
@@ -227,7 +276,7 @@ export class PixiLayerCustomData extends AbstractLayer {
    * @param styleOverride Custom style
    */
   renderLines(frame, viewport, zoom, lines, styleOverride) {
-    const l10n = this.context.systems.l10n;
+    const l1n = this.context.systems.l1n;
```
Comment (bhousel) [inline modules/pixi/PixiLayerCustomData.js:279]: seems to be a typo in a bunch of places
```suggestion
    const l1n = this.context.systems.l10n;
```

**Review by bhousel — COMMENTED** [review]
I took a quick pass over this, and my initial impression is that it seems like you might be doing too much work?  

My mental model of how this might work is like:
- in `_hashchange`, take the `data` param and try to parse it as WKT.
- If that succeeds, you'll get some blob of GeoJSON
- call `this.setFile(geojson, '.geojson')`
- Done?  everything else in this file is already set up to deal with GeoJSON.

**Comment by Bonkles** [comment]
> I took a quick pass over this, and my initial impression is that it seems like you might be doing too much work?
> _<snip>_
> ... everything else in this file is already set up to deal with GeoJSON.

Excellent idea. Refactoring to do _exactly_ this!

> in _hashchange, take the data param and try to parse it as WKT.

One downside: printing appropriate error messages with the overloaded `data` param. I can still wire up the separate `wktPoly` param and make it use the same code path of `this._setFile(geojson, '.geojson')`, or I can force err'thang to use 'data'. What do you think?

**Comment by bhousel** [comment]
> One downside: printing appropriate error messages with the overloaded `data` param. I can still wire up the separate `wktPoly` param and make it use the same code path of `this._setFile(geojson, '.geojson')`, or I can force err'thang to use 'data'. What do you think?

I'd force everything to just use `data=`   (for now)
I can imagine a situation where people want to see a bunch of custom data sources on the screen, e.g. an area of interest and PMTiles, but we aren't set up to support multiple custom data or imagery sources yet (but we should add that later!)

Good callout on the downside of not being able to print appropriate error messages - I think that's probably another thing we don't do at all right now? e.g. if you pass it a bad geojson or PMTiles, I think it just silently fails today?  We should probably eventually surface this somehow.

**Comment by bhousel** [comment]
Looks good! I do think we can simplify further and remove the arbitrary limitation to only include Polygon/Multipolygon types, but let's merge it so people can try it out.

**Pull request merged**